source/common/network/raw_buffer_socket.cc

```diff
@@ -19,7 +19,7 @@ IoResult RawBufferSocket::doRead(Buffer::Instance& buffer) {
   bool end_stream = false;
   do {
     // 16K read is arbitrary. TODO(mattklein123) PERF: Tune the read size.
-    Api::IoCallUint64Result result = callbacks_->ioHandle().read(buffer, 16384);
+    Api::IoCallUint64Result result = callbacks_->ioHandle().read(buffer, 131072);
```

Comment:
That's a good point. The existing code is wrong for really low limits, but that's a lot less likely to happen. Docs recommend 32k for edge deployments now. I'll try to come up with something here.

Reply:
Hmm, from this comment, I think the watermark buffer limits will be off-by-one in a way that will be detrimental to performance. So either the WatermarkBuffer needs to know the off-by-one value, or the caller of `Buffer::reserve` should check the connection `bufferLimit()` and not try to read more than that. Any thoughts on this @alyssawilk and how to make all the pieces play nicely with each other?

```
void ConnectionImpl::setBufferLimits(uint32_t limit) {
  read_buffer_limit_ = limit;

  // Due to the fact that writes to the connection and flushing data from the connection are done
  // asynchronously, we have the option of either setting the watermarks aggressively, and regularly
  // enabling/disabling reads from the socket, or allowing more data, but then not triggering
  // based on watermarks until 2x the data is buffered in the common case. Given these are all soft
  // limits we err on the side of buffering more triggering watermark callbacks less often.
  //
  // Given the current implementation for straight up TCP proxying, the common case is reading
  // |limit| bytes through the socket, passing |limit| bytes to the connection (triggering the high
  // watermarks) and the immediately draining |limit| bytes to the socket (triggering the low
  // watermarks). We avoid this by setting the high watermark to limit + 1 so a single read will
  // not trigger watermarks if the socket is not blocked.
  //
  // If the connection class is changed to write to the buffer and flush to the socket in the same
  // stack then instead of checking watermarks after the write and again after the flush it can
  // check once after both operations complete. At that point it would be better to change the high
  // watermark from |limit + 1| to |limit| as the common case (move |limit| bytes, flush |limit|
  // bytes) would not trigger watermarks but a blocked socket (move |limit| bytes, flush 0 bytes)
  // would result in respecting the exact buffer limit.
  if (limit > 0) {
    static_cast<Buffer::WatermarkBuffer*>(write_buffer_.get())->setWatermarks(limit + 1);
    read_buffer_.setWatermarks(limit + 1);
  }
}
```

Reply:
Also, I don't know that the `bufferLimit()` has any effect on the size of reads. The transport socket determines how much it wants to read. Maybe that was not the case when this was written (which I think predates transport sockets)?

Reply:
So that +1 in the read buffer limit is an optimization.  See this comment in a PR under review: https://github.com/envoyproxy/envoy/pull/14043#discussion_r525628234

And actually, the buffer high watermark has that +1 build in internally since the high-watermark computation uses > instead of >= to determine if the high-watermark has triggered.

Buffer limit does not have an effect on read size today, but we do rely on read size being 16KB to not go way over.